pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+//#[cfg(feature = "runtime-benchmarks")]
+//mod benchmarking;
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{Currency, ExistenceRequirement, ReservableCurrency},
+	transactional, PalletId,
+};
+use frame_system::pallet_prelude::*;
+use num_traits::{NumAssignOps, NumOps, Signed};
+use sp_runtime::{
+	traits::{AccountIdConversion, BlockNumberProvider, Saturating, Zero},
+	FixedPointNumber, FixedPointOperand, TokenError,
+};
+
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct EpochDetails<BlockNumber, Balance> {
+	ends_on: BlockNumber,
+	total_reward: Balance,
+}
+
+/// Type used to initialize the first epoch with the correct block number
+pub struct FirstEpochDetails<P>(std::marker::PhantomData<P>);
+impl<P, N, B: Zero> Get<EpochDetails<N, B>> for FirstEpochDetails<P>
+where
+	P: BlockNumberProvider<BlockNumber = N>,
+{
+	fn get() -> EpochDetails<N, B> {
+		EpochDetails {
+			ends_on: P::current_block_number(),
```

Comment:
This means that the `on_initialize` will run in the same block when the code is enacted. Can we add a bit of space between there? Like static `100` blocks

Reply:
This is the epoch that will end when the code is enacted. In the first `on_initialize` a new `EpochDetails` will be created using the correct `NextTotalReward`.
It can be understood as at the beginning of time, just a new epoch starts, and this code forces this to occur.

Reply:
I understand that, I am asking that we add a delay  of `n` blocks so it doesnt occur at the same time as the runtime upgrade first block.